frame/evm/src/runner/native.rs

```diff
@@ -0,0 +1,718 @@
+				source,
+				address,
+				value,
+				init,
+				Vec::new()
+			);
+
+			let mut create_info = CreateInfo {
+				exit_reason: reason.into(),
+				value: address,
+			};
+
+			match reason {
+				ExitReason::Succeed(_) => {
+					match substate.gasometer.record_deposit(out.len()) {
+						Ok(()) => {
+							AccountCodes::insert(address, out);
+							TransactionOutcome::Commit(Ok(create_info))
+						},
+						Err(e) => {
+							create_info.exit_reason = e.into();
+							TransactionOutcome::Rollback(Ok(create_info))
+						},
+					}
+				},
+				ExitReason::Revert(_) => TransactionOutcome::Rollback(Ok(create_info)),
+				ExitReason::Error(_) => TransactionOutcome::Rollback(Ok(create_info)),
+				ExitReason::Fatal(_) => TransactionOutcome::Rollback(Ok(create_info)),
+			}
+		})
+	}
+}
+
+fn l64(gas: usize) -> usize {
+	gas - gas / 64
+}
+
+pub struct Handler<'vicinity, 'config, T> {
+	vicinity: &'vicinity Vicinity,
+	config: &'config Config,
+	gasometer: Gasometer<'config>,
+	deleted: BTreeSet<H160>,
+	precompile: fn(H160, &[u8], Option<usize>) ->
+		Option<Result<(ExitSucceed, Vec<u8>, usize), ExitError>>,
+	is_static: bool,
+	_marker: PhantomData<T>,
+}
+
+impl<'vicinity, 'config, T: Trait> Handler<'vicinity, 'config, T> {
+	/// Create a new handler with given vicinity.
+	pub fn new_with_precompile(
+		vicinity: &'vicinity Vicinity,
+		gas_limit: usize,
+		is_static: bool,
+		config: &'config Config,
+		precompile: fn(H160, &[u8], Option<usize>) ->
+			Option<Result<(ExitSucceed, Vec<u8>, usize), ExitError>>,
+	) -> Self {
+		Self {
+			vicinity,
+			config,
+			is_static,
+			gasometer: Gasometer::new(gas_limit, config),
+			precompile,
+			deleted: BTreeSet::default(),
+			_marker: PhantomData,
+		}
+	}
+
+	pub fn execute(
+		&mut self,
+		caller: H160,
+		address: H160,
+		value: U256,
+		code: Vec<u8>,
+		input: Vec<u8>,
+	) -> (ExitReason, Vec<u8>) {
+		let context = Context {
+			caller,
+			address,
+			apparent_value: value,
+		};
+
+		let mut runtime = Runtime::new(
+			Rc::new(code),
+			Rc::new(input),
+			context,
+			self.config,
+		);
+
+		let reason = match runtime.run(self) {
+			Capture::Exit(s) => s,
+			Capture::Trap(_) => unreachable!("Trap is Infallible"),
+		};
+
+		match reason {
+			ExitReason::Succeed(s) => {
+				(s.into(), runtime.machine().return_value())
+			},
+			ExitReason::Error(e) => {
+				(e.into(), Vec::new())
+			},
+			ExitReason::Revert(e) => {
+				(e.into(), runtime.machine().return_value())
+			},
+			ExitReason::Fatal(e) => {
+				self.gasometer.fail();
+				(e.into(), Vec::new())
+			},
+		}
+	}
+
+	fn transfer(&self, transfer: Transfer) -> Result<(), ExitError> {
+		let source = T::AddressMapping::into_account_id(transfer.source);
+		let target = T::AddressMapping::into_account_id(transfer.target);
+
+		T::Currency::transfer(
+			&source,
+			&target,
+			transfer.value.low_u128().unique_saturated_into(),
+			ExistenceRequirement::AllowDeath,
+		).map_err(|_| ExitError::OutOfGas)
```

Comment:
why `OutOfGas`?

Reply:
If the transfer fails we need to early return an error. In EVM it doesn't matter much what error we return here. The usual choice for a blanket error is `OutOfGas`, so here it is. We can also change this to `ExitError::Other(_)` and return a string here.